### PR TITLE
Add more store flags

### PR DESCRIPTION
**Pull request by craigfurman**
* [x] I added CHANGELOG entry for this change.
* [ ] Change is not relevant to the end user.

## Changes

Add more flags to thanos-store: max-time, min-time, and store.grpc.series-sample-limit.

Implements https://github.com/thanos-io/kube-thanos/issues/175

## Verification

`make generate validate`, and I've experimented with it in a consumer repository that I use to manage thanos.

### CODE REVIEW

File: jsonnet/kube-thanos/kube-thanos-store-default-params.libsonnet [hunk 1]
```diff
@@ -12,7 +12,10 @@
   ignoreDeletionMarksDelay: '24h',
   logLevel: 'info',
   logFormat: 'logfmt',
+  maxTime: null,
+  minTime: null,
   resources: {},
+  storeGRPCSeriesSampleLimit: null,
```
Comment (craigfurman) [inline jsonnet/kube-thanos/kube-thanos-store-default-params.libsonnet:18]: Putting nulls here has a few benefits IMO:

- it doesn't duplicate upstream defaults
- it documents the existence of these keys, and how they're formulated compared to the corresponding upstream flags (e.g. camelCase vs kebab-case)
- it assists the implementation of `util.withOptionalArgs` (see thread below)

File: jsonnet/kube-thanos/kube-thanos-store.libsonnet [hunk 2]
```diff
@@ -78,7 +79,11 @@ function(params) {
             { config+: { service_name: defaults.name } } + ts.config.tracing
           ),
         ] else []
-      ),
+      ) + util.withOptionalArgs(ts.config, {
+        'max-time': 'maxTime',
+        'min-time': 'minTime',
+        'store.grpc.series-sample-limit': 'storeGRPCSeriesSampleLimit',
+      }),
```
Comment (metalmatze) [inline jsonnet/kube-thanos/kube-thanos-store.libsonnet:86]: I think it should be possible to add these flags only of they are not null.
Something along the lines of this:
```jsonnet
     args: [
       ...
      ] + (
        if ts.config.maxTime != null then ['--max-time=' + ts.config.maxTime] else []
      ) + (
```

What this does is either merge `[]` to the `args` array or merge an array with the flag `['--max-time=foo']`. The first case doesn't change the args array and is therefore a no-op.

File: jsonnet/kube-thanos/util.libsonnet [hunk 3]
```diff
@@ -0,0 +1,8 @@
+{
+  withOptionalArgs(config, flagsToKeysMap)::
+    [
+      '--%s=%s' % [flag, config[flagsToKeysMap[flag]]]
+      for flag in std.objectFields(flagsToKeysMap)
+      if config[flagsToKeysMap[flag]] != null
+    ],
+}
```
Comment (metalmatze) [inline jsonnet/kube-thanos/util.libsonnet:8]: This should be needed in the end :blush:

**Review by metalmatze — CHANGES_REQUESTED** [review]

File: jsonnet/kube-thanos/kube-thanos-store.libsonnet (hunk 2 above)
Comment (craigfurman) [reply jsonnet/kube-thanos/kube-thanos-store.libsonnet:86]: I could add the 3 flags with 3 if-statements (and I see there are similar constructs in this file already) but that seems quite verbose to me 😅 

If we end up adding support for all thanos flags, each component will have a sea of these if-statements. I think we need something more general - and now that I look at the open PR list, it looks like the WIP in https://github.com/thanos-io/kube-thanos/pull/184 might be working towards that. 🤔

Comment (craigfurman) [reply jsonnet/kube-thanos/kube-thanos-store.libsonnet:86]: @metalmatze sorry for the impatience, but I don't want this to get stale if possible. what do you think of my comment above?

**Comment by kakkoyun** [comment]
@craigfurman Probably needs a rebase.

**Comment by craigfurman** [comment]
@kakkoyun thanks, rebased :+1: 

@metalmatze sorry for the prodding, but have you had a chance to read my reply on one of the threads above? https://github.com/thanos-io/kube-thanos/pull/187#discussion_r555236740

If you'd still prefer to remove that function and do the null checks inline, I can do that, but wanted to justify my initial decision first :slightly_smiling_face.

File: jsonnet/kube-thanos/util.libsonnet (hunk 3 above)
Comment (craigfurman) [reply jsonnet/kube-thanos/util.libsonnet:8]: Will wait for resolution to https://github.com/thanos-io/kube-thanos/pull/187/files#r555236740, and if you still want to remove this function I can do that :+1:

**Comment by craigfurman** [comment]
rebased again

**Comment by craigfurman** [comment]
In the time since I opened this, I see that minTime / maxTime have been implemented. I'll close this.

**Comment by kakkoyun** [comment]
@craigfurman Sorry that we've missed this 🙈  Thanks a million again for your contributions.

**Comment by craigfurman** [comment]
no worries! it happens.

**Pull request closed without merging**